# Рекомендация тарифов

## Описание проекта


Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Они хотят построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».

В распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы. Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных уже сделана.

Необходимо построить модель с максимально большим значением accuracy. Нужно довести долю правильных ответов по крайней мере до 0.75 и проверить accuracy на тестовой выборке.

**Путь к файлу:**
- /datasets/users_behavior.csv

**Описание данных таблицы:**
- Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:
- сalls — количество звонков,
- minutes — суммарная длительность звонков в минутах,
- messages — количество sms-сообщений,
- mb_used — израсходованный интернет-трафик в Мб,
- is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

**План действий:**

1. Ознакомление с данными.
2. Разделить исходные данные на обучающую, валидационную и тестовую выборки.
3. Исследование качество разных моделей, меняя гиперпараметры.
4. Проверка качество модели на тестовой выборке.
5. Дополнительно: проверка модели на вменяемость.

### Глава 1. Открытие файла с данными и изучение общей информации

In [1]:
# импорт библиотек
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats as st

from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split


pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
# чтение файла с данными 
try:
    df = pd.read_csv('/datasets/users_behavior.csv')
except:
    df = pd.read_csv('users_behavior.csv')

In [3]:
# просмотр и изучение данных
df.info()
display(df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
5      58.0   344.56      21.0  15823.37         0
6      57.0   431.64      20.0   3738.90         1
7      15.0   132.40       6.0  21911.60         0
8       7.0    43.39       3.0   2538.67         1
9      90.0   665.41      38.0  17358.61         0
10     82.0   560.51      20.0   9619.53         1
11     45.0   344.32      13.0  19898.81         0
12     51.0   437.13      61.0  21523.58         0
13     56.0   433.07      16.0  16702.36         0
14    108.0   587.90       0.0  14406.50         1
15      6.0    22.13       0.0   2710.09         0
16      2.0    18.73       0.0    588.89         0
17     26.0   163.62       4.0  16870.34         0
18     79.0   532.62      90.0  19908.31         0
19     49.0   341.67      81.0  11770.28         1
20     93.0   691.37      31.0  25950.28         1
21     56.0   411.46      97.0  15393.91         0
22     48.0   351.60      66.0  16469.90         1
23     11.0    96.05      19.0   2305.74         0
24     56.0   360.30      30.0  13560.15         0
25     53.0   356.22      52.0  22471.77         0
26     81.0   624.71      38.0   8644.66         1
27    154.0  1078.64      48.0  29335.15         1
28     37.0   286.68      44.0  19542.45         0
29     50.0   358.01      44.0   9998.83         1
30     41.0   276.81       0.0  19815.62         0
31     10.0    77.09      31.0    643.15         0
32     71.0   530.13      27.0  20698.51         0
33     65.0   482.35       7.0  23064.58         1
34    110.0   776.85       9.0  20790.93         1
35    120.0   814.85       0.0  20427.00         1
36     76.0   543.18      43.0  31845.11         1
37     64.0   422.69       0.0  11460.29         0
38     23.0   146.18      33.0   9711.45         0
39     81.0   519.48      34.0  12750.82         0
40     34.0   183.28       0.0  18631.66         0
41     11.0    85.31       7.0  10764.82         1
42     48.0   392.04       0.0  19122.13         0
43     50.0   323.99      10.0  20922.05         1
44     98.0   665.56     106.0  19538.83         0
45     35.0   235.58      97.0  20710.71         0
46     76.0   535.91      65.0  11968.22         0
47     49.0   327.45       0.0  12049.91         1
48      5.0    49.95      17.0   3370.69         1
49     58.0   474.92      11.0  17858.59         0
50     70.0   532.10      29.0  12403.26         0
51    124.0   842.40      75.0  21544.34         1
52    129.0   929.23       0.0  22508.96         1
53     67.0   421.27      33.0  14865.74         1
54      0.0     0.00      33.0  14010.33         1
55     13.0   106.03      16.0  37328.45         1
56     68.0   449.29      47.0  18171.20         0
57     91.0   585.09      49.0  27339.85         1
58     77.0   477.54       0.0  19090.74         1
59     51.0   356.39      33.0  20456.97         0
60    121.0   853.01      64.0  26605.94         1
61    114.0   823.58     144.0  30323.43         1
62     93.0   622.66       0.0  20894.56         1
63    125.0   814.53      61.0  16917.44         1
64     67.0   390.62      99.0  12991.39         0
65     98.0   677.18      46.0  16285.67         0
66     80.0   556.36      64.0  24049.46         0
67     33.0   238.70      24.0   9308.63         1
68    138.0  1009.11      64.0  27807.13         1
69     84.0   607.05      22.0  24875.03         0
70     77.0   546.20      55.0  11457.44         0
71     78.0   510.34       7.0  21765.50         0
72     69.0   477.53       7.0  21437.29         0
73     58.0   392.14       8.0  22525.18         0
74     63.0   427.64      93.0  19541.17         0
75     72.0   531.80       0.0  11299.60         0
76     73.0   633.22      21.0  18421.17         0
77      1.0     1.00     153

5 столбцов, 3214 строк, нулевых строк нет

### Глава 2. Разделение исходные данные на обучающую, валидационную и тестовую выборки.

In [4]:
features = df.drop('is_ultra', axis=1) # извлекаем признаки
target = df['is_ultra'] # извлекаем целевой признак

 
# Разделение исходных данных на обучающую и валидационную и тестовую выборки
features_train_valid, features_test, target_train_valid, target_test = train_test_split(
    features, target, stratify=target, test_size=0.25, random_state=12345)

# Далее разделяем обучающую и валидационную выборку на отдельные обучающую (features_train, target_train) и валидационную (features_valid, target_valid) выборки
features_train, features_valid, target_train, target_valid = train_test_split(
    features_train_valid, target_train_valid, stratify=target_train_valid, test_size=0.25, random_state=12345)


In [20]:
display(features_test.shape)
target_test.value_counts()

(804, 4)

0    558
1    246
Name: is_ultra, dtype: int64

In [21]:
display(features_train.shape)
target_train.value_counts()

(1807, 4)

0    1253
1     554
Name: is_ultra, dtype: int64

In [22]:
display(features_valid.shape)
target_valid.value_counts()

(603, 4)

0    418
1    185
Name: is_ultra, dtype: int64

In [23]:
#!pip install fast_ml

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.1/42.1 kB 668.6 kB/s eta 0:00:00 0:00:01


In [24]:
#from fast_ml.model_development import train_valid_test_split
#X_train, y_train, X_valid, y_valid, X_test, y_test = train_valid_test_split(df, target='is_ultra',
                                                                           train_size=0.6, valid_size=0.2, test_size=0,2)

### Глава 3. Исследование качества разных моделей

*Данная задача относиться к задачам классификации, в нашем случае бинарной классификации.*

Рассмотрим 2 модели:

- Логистическая регрессия LogisticRegression из библиотеки sklearn.linear_model
- Случайный лес RandomForestClassifier из библиотеки sklearn.ensemble

Сразу введем псевдослучайность для алгоритма обучения, используя генератор псевдослучайных чисел random_state. Таким образом модель будет воспринимать данные как случайные, для нас это необходимо , чтобы результаты неизменно получались одинаковыми (другими словами, повторить удачный эксперимент).

### LogisticRegression

In [5]:
from sklearn.linear_model import LogisticRegression
from joblib import dump


model_lr = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=300)
model_lr.fit(features_train, target_train) #скармливаем параметры

# Оценка качества модели на обучающей выборке
train_accuracy_lr = model_lr.score(features_train, target_train)
print('Accuracy тренирочоной модели:', train_accuracy_lr) #выводим accuracy тренировочной модели 

# Оценка качества модели на валидационной выборке
valid_accuracy_lr = model_lr.score(features_valid, target_valid)
print('Accuracy валидационной модели:',valid_accuracy_lr) #выводим accuracy валидационной модели

# Сохраним модель для дальнейшего использования
dump(model_lr, 'model_log_reg.joblib')


Accuracy тренирочоной модели: 0.7487548422800221
Accuracy валидационной модели: 0.7412935323383084


['model_log_reg.joblib']

### Случайный лес

In [6]:
from sklearn.ensemble import RandomForestClassifier

best_model_rf = None
best_result_accuracy_rf = 0

# Создание и обучение модели случайного леса с алгоритмом поиска лучшего количества оценщиков
for est in range(1, 10):
    model_rf = RandomForestClassifier(random_state=12345, n_estimators=est) # обучим модель с заданным количеством деревьев
    model_rf.fit(features_train, target_train) # обучим модель на тренировочной выборке
    result_accuracy_rf = model_rf.score(features_valid, target_valid) # посчитайте качество модели на валидационной выборке
    if result_accuracy_rf > best_result_accuracy_rf:
        best_model_rf = model_rf # сохраните наилучшую модель
        best_result_accuracy_rf = result_accuracy_rf #  сохраните наилучшее значение метрики accuracy на валидационных данных
        
        

print("Accuracy наилучшей модели на валидационной выборке:", best_result_accuracy_rf)

# Сохраним модель для дальнейшего использования
dump(best_model_rf, 'model_random_forest.joblib')

Accuracy наилучшей модели на валидационной выборке: 0.7993366500829188


['model_random_forest.joblib']

## Проверим модель на тестовой выборке

In [8]:
# Оценка качества модели логистической регресс на тестовой выборке
test_accuracy_lr = model_lr.score(features_test, target_test)
print('Accuracy тестовой модели (логистической регрессии):', test_accuracy_lr)

# Оценка качества модели случайного леса на тестовой выборке
test_accuracy_rf = model_rf.score(features_test, target_test)
print('Accuracy тестовой модели (случайный лес):', test_accuracy_rf)


Accuracy тестовой модели (логистической регрессии): 0.7611940298507462
Accuracy тестовой модели (случайный лес): 0.7723880597014925


## Вывод

**Модель Случайный лес показала более хорошие результаты из 2 моделей.**
- Полученные значения качества модели в 77% говорит о хорошем предсказании.
- Построенная модель для задачи классификации, сможет подобрать для пользователей, пользующихся архивными тарифами, более подходящий тариф («Смарт» или «Ультра»).
- Проанализировав поведение клиентов: тратах минут, сообщений и интернет трафика, модель в 4 из 5 случаев предскажет верный тариф для пользователя.
- Модель подходит как система "Рекомендации тарифов"